src/cost_estimates/fee_medians.rs

```diff
@@ -0,0 +1,348 @@
+        tx.execute(
+            insert_sql,
+            rusqlite::params![new_measure.high, new_measure.middle, new_measure.low,],
+        )
+        .expect("SQLite failure");
+        tx.execute(deletion_sql, rusqlite::params![self.window_size])
+            .expect("SQLite failure");
+
+        let estimate = Self::get_rate_estimates_from_sql(&tx, self.window_size);
+        tx.commit().expect("SQLite failure");
+        if let Ok(next_estimate) = estimate {
+            debug!("Updating fee rate estimate for new block";
+                   "new_measure_high" => new_measure.high,
+                   "new_measure_middle" => new_measure.middle,
+                   "new_measure_low" => new_measure.low,
+                   "new_estimate_high" => next_estimate.high,
+                   "new_estimate_middle" => next_estimate.middle,
+                   "new_estimate_low" => next_estimate.low);
+        }
+    }
+}
+
+impl<M: CostMetric> FeeEstimator for WeightedMedianFeeRateEstimator<M> {
+    fn notify_block(
+        &mut self,
+        receipt: &StacksEpochReceipt,
+        block_limit: &ExecutionCost,
+    ) -> Result<(), EstimatorError> {
+        // Calculate sorted fee rate for each transaction in the block.
+        let mut working_fee_rates: Vec<FeeRateAndWeight> = receipt
+            .tx_receipts
+            .iter()
+            .filter_map(|tx_receipt| {
+                fee_rate_and_weight_from_receipt(&self.metric, &tx_receipt, block_limit)
+            })
+            .collect();
+
+        // If necessary, add the "minimum" fee rate to fill the block.
+        maybe_add_minimum_fee_rate(&mut working_fee_rates, self.full_block_weight);
+
+        // If fee rates non-empty, then compute an update.
+        if working_fee_rates.len() > 0 {
+            // Values must be sorted.
+            working_fee_rates.sort_by(|a, b| {
+                a.fee_rate
+                    .partial_cmp(&b.fee_rate)
+                    .unwrap_or(Ordering::Equal)
+            });
+
+            // Compute the estimate and update.
+            let block_estimate = fee_rate_estimate_from_sorted_weighted_fees(&working_fee_rates);
+            self.update_estimate(block_estimate);
+        }
+
+        Ok(())
+    }
+
+    fn get_rate_estimates(&self) -> Result<FeeRateEstimate, EstimatorError> {
+        Self::get_rate_estimates_from_sql(&self.db, self.window_size)
+    }
+}
+
+/// Computes a `FeeRateEstimate` based on `sorted_fee_rates` using a "weighted percentile" method
+/// described in https://en.wikipedia.org/wiki/Percentile#Weighted_percentile
+///
+/// The percentiles computed are [0.05, 0.5, 0.95].
+///
+/// `sorted_fee_rates` must be non-empty.
+fn fee_rate_estimate_from_sorted_weighted_fees(
+    sorted_fee_rates: &Vec<FeeRateAndWeight>,
+) -> FeeRateEstimate {
+    if sorted_fee_rates.is_empty() {
+        panic!("`sorted_fee_rates` cannot be empty.");
+    }
+
+    let mut total_weight = 0u64;
```

Comment:
You can avoid worrying about overflows in `total_weight` and `cumulative_weight` by using `f64` for them (especially because they get cast to f64 anyways)

Reply:
done in latest